File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] left = new int[n];
        int[] right = new int[n];
        // store the left
        for (int i = 0; i < n; i++) {
            if (i == 0 || left[i - 1] < height[i])
                left[i] = height[i];
            else {
                left[i] = left[i - 1];
            }
        }
        // store the right
        for (int i = n - 1; i >= 0; i--) {
            if (i == n - 1 || right[i + 1] < height[i])
                right[i] = height[i];
            else {
                right[i] = right[i + 1];
            }
        }
        System.out.println("left=" + Arrays.toString(left));
        System.out.println("right=" + Arrays.toString(right));

        int water = 0;
        for (int i = 0; i < n; i++) {
            int h = height[i];
            water += Math.min(left[i], right[i]) - h;
        }
        return water;

    }
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def tapping_rain_water2(self, arr):
        n=len(arr)
        left_walls=[0 for _ in range(n)]
        left_walls[0] = arr[0]
        # traverse from left to right and store the immdediate bigger buildings
        for i in range(1,n):
            left_walls[i] = max(arr[i], left_walls[i-1])
        
        right_walls=[0 for _ in range(n)]
        right_walls[n-1] = arr[n-1]
        # traverse from right to left and store the immdediate bigger builders
        for i in range(n-2, -1, -1):
            right_walls[i] = max(arr[i], right_walls[i+1])

        # result
        water=0
        # traverse and keep building and calc the water
        for i in range(n):
            lefth=left_walls[i]
            righth=right_walls[i]
            
            water += min(lefth, righth) - arr[i]

        return water

    def trap(self, height: List[int]) -> int:
        return self.tapping_rain_water2(height)